File: zerodb-core/src/frontier.rs
//! Causal frontiers and snapshot identity per doc/FRONTIER.md (M0f).

use std::collections::BTreeMap;

use crate::merkle::{MerkleOp, merkle_root};

pub const DOMAIN_SNAPSHOT: &[u8] = b"zerodb-snapshot-v1";
pub const SNAPSHOT_FORMAT_VERSION: u8 = 1;

/// One author's tip: greatest accepted op under KERNEL §4.5, with the
/// HLC needed to decide late-op from the encoded frontier alone (CX-04).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct FrontierTip {
    pub op_id: [u8; 32],
    pub physical_ms: u64,
    pub logical: u16,
}

/// Frontier: PeerId (author) → tip.
pub type Frontier = BTreeMap<[u8; 32], FrontierTip>;

fn order_key(op: &MerkleOp) -> (u64, u16, [u8; 32], [u8; 32]) {
    (op.physical_ms, op.logical, op.author, op.op_id)
}

/// Build frontier from accepted ops.
pub fn build_frontier(ops: &[MerkleOp]) -> Frontier {
    let mut best: BTreeMap<[u8; 32], &MerkleOp> = BTreeMap::new();
    for op in ops {
        best.entry(op.author)
            .and_modify(|cur| {
                if order_key(op) > order_key(cur) {
                    *cur = op;
                }
            })
            .or_insert(op);
    }
    best.into_iter()
        .map(|(author, op)| {
            (
                author,
                FrontierTip {
                    op_id: op.op_id,
                    physical_ms: op.physical_ms,
                    logical: op.logical,
                },
            )
        })
        .collect()
}

/// Tail boundary: greatest OpId under full §4.5 order, or None if empty.
pub fn tail_boundary(ops: &[MerkleOp]) -> Option<[u8; 32]> {
    ops.iter()
        .max_by(|a, b| order_key(a).cmp(&order_key(b)))
        .map(|o| o.op_id)
}

/// Canonical frontier encoding: sorted by author, each
/// author ‖ op_id ‖ physical_ms (u64 BE) ‖ logical (u16 BE).
pub fn frontier_bytes(f: &Frontier) -> Vec<u8> {
    let mut out = Vec::new();
    for (author, tip) in f {
        out.extend_from_slice(author);
        out.extend_from_slice(&tip.op_id);
        out.extend_from_slice(&tip.physical_ms.to_be_bytes());
        out.extend_from_slice(&tip.logical.to_be_bytes());
    }
    out
}

pub fn snapshot_id(
    datastore_id: &[u8; 32],
    frontier: &Frontier,
    merkle: &[u8; 32],
    tail: Option<&[u8; 32]>,
) -> [u8; 32] {
    let mut pre = Vec::new();
    pre.extend_from_slice(DOMAIN_SNAPSHOT);
    pre.push(SNAPSHOT_FORMAT_VERSION);
    pre.extend_from_slice(datastore_id);
    pre.extend_from_slice(&frontier_bytes(frontier));
    pre.extend_from_slice(merkle);
    match tail {
        Some(t) => {
            pre.push(1);
            pre.extend_from_slice(t);
        }
        None => pre.push(0),
    }
    *blake3::hash(&pre).as_bytes()
}

/// Late op: author's encoded tip already dominates this op in §4.5 order.
/// Uses only the frontier map — no extra op list (CX-04).
pub fn is_late_op(op: &MerkleOp, frontier: &Frontier) -> bool {
    match frontier.get(&op.author) {
        None => false,
        Some(tip) => {
            let tip_key = (tip.physical_ms, tip.logical, op.author, tip.op_id);
            order_key(op) < tip_key && op.op_id != tip.op_id
        }
    }
}

/// Late check from a published op set: build the frontier, then `is_late_op`.
pub fn is_late_against_ops(op: &MerkleOp, frontier_ops: &[MerkleOp]) -> bool {
    is_late_op(op, &build_frontier(frontier_ops))
}

pub fn snapshot_for_ops(datastore_id: &[u8; 32], ops: &[MerkleOp]) -> [u8; 32] {
    let f = build_frontier(ops);
    let root = merkle_root(ops);
    let tail = tail_boundary(ops);
    snapshot_id(datastore_id, &f, &root, tail.as_ref())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn op(id: u8, author: u8, p: u64) -> MerkleOp {
        MerkleOp {
            op_id: [id; 32],
            physical_ms: p,
            logical: 0,
            author: [author; 32],
        }
    }

    #[test]
    fn frontier_per_author_max() {
        let ops = vec![op(1, 0xAA, 100), op(2, 0xAA, 200), op(3, 0xBB, 150)];
        let f = build_frontier(&ops);
        assert_eq!(f.get(&[0xAA; 32]).map(|t| t.op_id), Some([2; 32]));
        assert_eq!(f.get(&[0xBB; 32]).map(|t| t.op_id), Some([3; 32]));
        assert_eq!(f.get(&[0xAA; 32]).map(|t| t.physical_ms), Some(200));
    }

    #[test]
    fn late_op_detected() {
        let base = vec![op(2, 0xAA, 200)];
        let late = op(1, 0xAA, 100);
        let f = build_frontier(&base);
        assert!(is_late_op(&late, &f));
        assert!(!is_late_op(&op(3, 0xAA, 300), &f));
        assert!(is_late_against_ops(&late, &base));
        assert!(!is_late_against_ops(&op(3, 0xAA, 300), &base));
    }

    #[test]
    fn snapshot_stable() {
        let ds = [0xDDu8; 32];
        let ops = vec![op(1, 0xAA, 1000)];
        assert_eq!(snapshot_for_ops(&ds, &ops), snapshot_for_ops(&ds, &ops));
    }

    #[test]
    #[ignore]
    fn print_snapshot_golden() {
        fn hex(b: &[u8]) -> String {
            b.iter().map(|x| format!("{x:02x}")).collect()
        }
        let ds = [0xDDu8; 32];
        let ops = vec![op(1, 0xAA, 1000)];
        eprintln!("snap={}", hex(&snapshot_for_ops(&ds, &ops)));
        eprintln!("root={}", hex(&merkle_root(&ops)));
        let f = build_frontier(&ops);
        eprintln!("frontier_author={}", hex(&[0xAAu8; 32]));
        eprintln!("frontier_tip={}", hex(&f.get(&[0xAAu8; 32]).unwrap().op_id));
    }
}

File: conformance/ts/models/frontier.mjs
// Frontiers + snapshot identity per doc/FRONTIER.md (M0f).

import { blake3 } from './blake3.mjs';
import { bytesToHex, hexToBytes } from './cbor.mjs';
import { merkleRootOnce } from './merkle.mjs';

const DOMAIN_SNAPSHOT = new TextEncoder().encode('zerodb-snapshot-v1');
const SNAPSHOT_FORMAT_VERSION = 1;

function parseOp(o) {
  return {
    op_id: hexToBytes(o.op_id),
    physical_ms: o.physical_ms,
    logical: o.logical ?? 0,
    author: hexToBytes(o.author),
  };
}

function orderKey(op) {
  return [op.physical_ms, op.logical, bytesToHex(op.author), bytesToHex(op.op_id)];
}

function cmpKey(a, b) {
  const ka = orderKey(a);
  const kb = orderKey(b);
  for (let i = 0; i < ka.length; i++) {
    if (ka[i] < kb[i]) return -1;
    if (ka[i] > kb[i]) return 1;
  }
  return 0;
}

export function buildFrontier(ops) {
  const best = new Map();
  for (const op of ops) {
    const a = bytesToHex(op.author);
    if (!best.has(a) || cmpKey(op, best.get(a)) > 0) best.set(a, op);
  }
  const out = {};
  for (const [a, op] of [...best.entries()].sort((x, y) => (x[0] < y[0] ? -1 : 1))) {
    out[a] = {
      op_id: bytesToHex(op.op_id),
      physical_ms: op.physical_ms,
      logical: op.logical,
    };
  }
  return out;
}

export function tailBoundary(ops) {
  if (ops.length === 0) return null;
  let best = ops[0];
  for (const op of ops) if (cmpKey(op, best) > 0) best = op;
  return bytesToHex(best.op_id);
}

function frontierBytes(frontier) {
  const authors = Object.keys(frontier).sort();
  const parts = [];
  for (const a of authors) {
    const tip = frontier[a];
    parts.push(hexToBytes(a));
    parts.push(hexToBytes(typeof tip === 'string' ? tip : tip.op_id));
    const p = typeof tip === 'string' ? 0 : tip.physical_ms;
    const l = typeof tip === 'string' ? 0 : tip.logical;
    const pb = new Uint8Array(8);
    new DataView(pb.buffer).setBigUint64(0, BigInt(p), false);
    const lb = new Uint8Array(2);
    new DataView(lb.buffer).setUint16(0, l, false);
    parts.push(pb);
    parts.push(lb);
  }
  let n = 0;
  for (const p of parts) n += p.length;
  const out = new Uint8Array(n);
  let o = 0;
  for (const p of parts) {
    out.set(p, o);
    o += p.length;
  }
  return out;
}

export function snapshotId(datastoreIdHex, frontier, merkleHex, tailHex) {
  const parts = [
    DOMAIN_SNAPSHOT,
    new Uint8Array([SNAPSHOT_FORMAT_VERSION]),
    hexToBytes(datastoreIdHex),
    frontierBytes(frontier),
    hexToBytes(merkleHex),
  ];
  if (tailHex) {
    parts.push(new Uint8Array([1]));
    parts.push(hexToBytes(tailHex));
  } else {
    parts.push(new Uint8Array([0]));
  }
  let n = 0;
  for (const p of parts) n += p.length;
  const pre = new Uint8Array(n);
  let o = 0;
  for (const p of parts) {
    pre.set(p, o);
    o += p.length;
  }
  return bytesToHex(blake3(pre));
}

export function isLateOp(op, frontier) {
  const tip = frontier[bytesToHex(op.author)];
  if (!tip) return false;
  const tipOp = {
    op_id: hexToBytes(typeof tip === 'string' ? tip : tip.op_id),
    physical_ms: typeof tip === 'string' ? 0 : tip.physical_ms,
    logical: typeof tip === 'string' ? 0 : tip.logical,
    author: op.author,
  };
  return cmpKey(op, tipOp) < 0 && bytesToHex(op.op_id) !== bytesToHex(tipOp.op_id);
}

export function isLateAgainstOps(op, frontierOps) {
  return isLateOp(op, buildFrontier(frontierOps));
}

export function runFrontierBuildVector(vector) {
  const ops = (vector.ops || []).map(parseOp);
  const got = buildFrontier(ops);
  const gotIds = {};
  for (const [a, tip] of Object.entries(got)) gotIds[a] = tip.op_id;
  if (JSON.stringify(gotIds) !== JSON.stringify(vector.expect_frontier)) {
    throw new Error(
      `frontier mismatch:\n  expected ${JSON.stringify(vector.expect_frontier)}\n  got ${JSON.stringify(gotIds)}`
    );
  }
}

export function runSnapshotIdVector(vector) {
  const ops = (vector.ops || []).map(parseOp);
  const f = buildFrontier(ops);
  const root = bytesToHex(merkleRootOnce(ops));
  const tail = tailBoundary(ops);
  const id = snapshotId(vector.datastore_id_hex, f, root, tail);
  if (id !== vector.expect_snapshot_id_hex) {
    throw new Error(`snapshot id: expected ${vector.expect_snapshot_id_hex}, got ${id}`);
  }
  if (vector.expect_merkle_root_hex && root !== vector.expect_merkle_root_hex) {
    throw new Error(`merkle root: expected ${vector.expect_merkle_root_hex}, got ${root}`);
  }
}

export function runLateOpVector(vector) {
  const op = parseOp(vector.op);
  const late = vector.encoded_frontier
    ? isLateOp(op, vector.encoded_frontier)
    : isLateAgainstOps(op, (vector.frontier_ops || []).map(parseOp));
  if (late !== vector.expect_late) {
    throw new Error(`late: expected ${vector.expect_late}, got ${late}`);
  }
}
